tests/helpers/test_script.py

```diff
@@ -15,63 +16,106 @@
+            self._timeout = timeout
+            self._loop = asyncio.get_event_loop()
+            self._task = None
+            self._cancelled = False
+            self._unsub = None
+
+        async def __aenter__(self):
+            if self._timeout is None:
+                return self
+            self._task = asyncio.Task.current_task()
+            if self._timeout <= 0:
+                self._loop.call_soon(self._cancel_task)
+                return self
+            # Wait for a time_changed event instead of real time passing.
+            self._unsub = async_call_later(hass, self._timeout, self._cancel_task)
+            return self
+
+        async def __aexit__(self, exc_type, exc_val, exc_tb):
+            if exc_type is asyncio.CancelledError and self._cancelled:
+                self._unsub = None
+                self._task = None
+                raise asyncio.TimeoutError
+            if self._timeout is not None and self._unsub:
+                self._unsub()
+                self._unsub = None
+            self._task = None
+            return None
 
+        @callback
+        def _cancel_task(self, now=None):
+            if self._task is not None:
+                self._task.cancel()
+                self._cancelled = True
 
-async def test_firing_event_basic(hass):
-    """Test the firing of events."""
-    event = "test_event"
-    context = Context()
+    monkeypatch.setattr(script, "timeout", MockTimeout)
 
-    @callback
-    def record_event(event):
-        """Add recorded event to set."""
-        events.append(event)
 
-    hass.bus.async_listen(event, record_event)
+def async_watch_for_action(script_obj, message):
+    """Watch for message in last_action."""
+    flag = asyncio.Event()
 
-    schema = cv.SCRIPT_SCHEMA({"event": event, "event_data": {"hello": "world"}})
+    @callback
+    def check_action():
+        if script_obj.last_action and message in script_obj.last_action:
+            flag.set()
 
-    # For this one test we'll make sure "legacy" works the same as None.
-    for run_mode in _ALL_RUN_MODES + ["legacy"]:
-        events = []
+    script_obj.change_listener = check_action
+    return flag
 
-        if run_mode is None:
-            script_obj = script.Script(hass, schema)
-        else:
-            script_obj = script.Script(hass, schema, run_mode=run_mode)
 
-        assert not script_obj.can_cancel
+@pytest.mark.parametrize("script_mode", _BASIC_SCRIPT_MODES)
+async def test_firing_event_basic(hass, script_mode):
+    """Test the firing of events."""
+    event = "test_event"
+    context = Context()
+    events = async_capture_events(hass, event)
 
-        await script_obj.async_run(context=context)
+    sequence = cv.SCRIPT_SCHEMA({"event": event, "event_data": {"hello": "world"}})
+    script_obj = script.Script(hass, sequence, script_mode=script_mode)
 
-        await hass.async_block_till_done()
+    assert script_obj.is_legacy == (script_mode == "legacy")
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5164426)